test/benchmarks.cpp

```diff
@@ -0,0 +1,460 @@
+// See LICENSE.md in the project root or https://www.gnu.org/licenses/gpl-3.0.en.html
+// for details.
+
+#include <atomic>
+#include <iostream>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "catch.hpp"
+#include "qfactory.hpp"
+
+#include "tests.hpp"
+
+using namespace Qrack;
+
+#define EPSILON 0.001
+#define REQUIRE_FLOAT(A, B) do {                                            \
+        double __tmp_a = A;                                                 \
+        double __tmp_b = B;                                                 \
+        REQUIRE(__tmp_a < (__tmp_b + EPSILON));                             \
+        REQUIRE(__tmp_b > (__tmp_b - EPSILON));                             \
+    } while (0);
+
+void benchmarkLoop(std::function<void(QInterfacePtr, int)> fn) {
+
+    const int ITERATIONS = 100;
+
+    std::cout<<std::endl;
+    std::cout<<ITERATIONS<<" iterations";
+    std::cout<<std::endl;
+    std::cout<<"# of Qubits, ";
+    std::cout<<"Average Time (ms), ";
+    std::cout<<"Sample Std. Deviation (ms), ";
+    std::cout<<"Fastest (ms), ";
+    std::cout<<"1st Quartile (ms), ";
+    std::cout<<"Median (ms)";
+    std::cout<<"3rd Quartile (ms), ";
+    std::cout<<"Slowest (ms)"<<std::endl;
+
+    clock_t startClock, tClock, allClock;
+    clock_t trialClocks[ITERATIONS];
+
+    int i, numBits;
+
+    double avgt, stdet;
+
+    // Grover's search inverts the function of a black box subroutine.
+    // Our subroutine returns true only for an input of 100.
+    for (numBits = 3; numBits <= 20; numBits++) {
+        QInterfacePtr qftReg = CreateQuantumInterface(testEngineType, testSubEngineType, numBits, 0, rng);
+        startClock = clock();
+        allClock = startClock;
+        tClock = startClock;
+        avgt = 0.0;
+        for (i = 0; i < ITERATIONS; i++) {
+
+            // Run loop body
+            fn(qftReg, numBits);
+
+            // Display timing
+            tClock = clock() - allClock;
+            trialClocks[i] = tClock;
+            avgt += tClock;
+            allClock = clock();
+        }
+        allClock = clock() - startClock;
+        avgt /= ITERATIONS;
+
+        stdet = 0.0;
+        for (i = 0; i < ITERATIONS; i++) {
+            stdet += (trialClocks[i] - avgt) * (trialClocks[i] - avgt);
+        }
+        stdet = sqrt(stdet / ITERATIONS);
+
+        std::sort(trialClocks, trialClocks + ITERATIONS);
+
+        std::cout<<(int)numBits<<", ";
+        std::cout<<(avgt * 1000.0 / CLOCKS_PER_SEC)<<",";
+        std::cout<<(stdet * 1000.0 / CLOCKS_PER_SEC)<<",";
+        std::cout<<(trialClocks[0] * 1000.0 / CLOCKS_PER_SEC)<<",";
+        std::cout<<(trialClocks[ITERATIONS / 4 - 1] * 1000.0 / CLOCKS_PER_SEC)<<",";
+        std::cout<<(trialClocks[ITERATIONS / 2 - 1] * 1000.0 / CLOCKS_PER_SEC)<<",";
+        std::cout<<(trialClocks[(3 * ITERATIONS) / 4 - 1] * 1000.0 / CLOCKS_PER_SEC)<<",";
+        std::cout<<(trialClocks[ITERATIONS - 1] * 1000.0 / CLOCKS_PER_SEC)<<std::endl;
```

Comment:
I really feel like calling this the “slowest” is kinda misrepresenting.  Wouldn’t it be more informative to consider these values on a per-qubit basis, on an exponential scale?

Reply:
How is this is a misrepresentation? The trial results are ordered from fastest to slowest iteration, and the quartile boundaries are reported, in addition to the average and standard deviation. They _are_ reported on a per qubit basis. Exponential scaling is the easiest part to do with graphing and analysis tools.

Reply:
I completely missed the sort. Code comprehension fail! Nevermind :)